File: web/lib/workflow-runtime/graph.ts
import type { WorkflowDSL, WorkflowNode } from '../workflow-types'

type GraphSnapshot = {
  nodeMap: Map<string, WorkflowNode>
  outgoing: Map<string, string[]>
  incomingCount: Map<string, number>
}

const cloneIncomingCount = (source: Map<string, number>) => {
  const next = new Map<string, number>()
  source.forEach((value, key) => next.set(key, value))
  return next
}

const buildGraphSnapshot = (dsl: WorkflowDSL): GraphSnapshot => {
  const nodeMap = new Map<string, WorkflowNode>()
  const outgoing = new Map<string, string[]>()
  const incomingCount = new Map<string, number>()

  dsl.nodes.forEach((node) => {
    nodeMap.set(node.id, node)
    outgoing.set(node.id, [])
    incomingCount.set(node.id, 0)
  })

  dsl.edges.forEach((edge) => {
    if (!nodeMap.has(edge.source) || !nodeMap.has(edge.target))
      return
    outgoing.set(edge.source, [...(outgoing.get(edge.source) ?? []), edge.target])
    incomingCount.set(edge.target, (incomingCount.get(edge.target) ?? 0) + 1)
  })

  return { nodeMap, outgoing, incomingCount }
}

export const buildExecutionPlan = (dsl: WorkflowDSL): string[] => {
  const graph = buildGraphSnapshot(dsl)
  const incoming = cloneIncomingCount(graph.incomingCount)
  const queue = [...dsl.nodes.filter(node => (incoming.get(node.id) ?? 0) === 0).map(node => node.id)]
  const ordered: string[] = []

  while (queue.length > 0) {
    const current = queue.shift()
    if (!current)
      continue
    ordered.push(current)
    const nextNodes = graph.outgoing.get(current) ?? []
    nextNodes.forEach((nextNodeId) => {
      const nextCount = (incoming.get(nextNodeId) ?? 0) - 1
      incoming.set(nextNodeId, nextCount)
      if (nextCount === 0)
        queue.push(nextNodeId)
    })
  }

  if (ordered.length !== dsl.nodes.length) {
    const missing = dsl.nodes.map(node => node.id).filter(nodeId => !ordered.includes(nodeId))
    return [...ordered, ...missing]
  }
  return ordered
}

File: server/internal/workflowruntime/graph.go
package workflowruntime

func BuildExecutionPlan(dsl WorkflowDSL) []string {
	nodeIDs := make([]string, 0, len(dsl.Nodes))
	for _, node := range dsl.Nodes {
		nodeIDs = append(nodeIDs, node.ID)
	}

	outgoing := make(map[string][]string, len(dsl.Nodes))
	incomingCount := make(map[string]int, len(dsl.Nodes))
	nodeExists := make(map[string]bool, len(dsl.Nodes))
	for _, node := range dsl.Nodes {
		outgoing[node.ID] = []string{}
		incomingCount[node.ID] = 0
		nodeExists[node.ID] = true
	}

	for _, edge := range dsl.Edges {
		if !nodeExists[edge.Source] || !nodeExists[edge.Target] {
			continue
		}
		outgoing[edge.Source] = append(outgoing[edge.Source], edge.Target)
		incomingCount[edge.Target] = incomingCount[edge.Target] + 1
	}

	queue := make([]string, 0)
	for _, nodeID := range nodeIDs {
		if incomingCount[nodeID] == 0 {
			queue = append(queue, nodeID)
		}
	}

	ordered := make([]string, 0, len(nodeIDs))
	incoming := make(map[string]int, len(incomingCount))
	for k, v := range incomingCount {
		incoming[k] = v
	}

	for len(queue) > 0 {
		current := queue[0]
		queue = queue[1:]
		ordered = append(ordered, current)
		for _, next := range outgoing[current] {
			incoming[next] = incoming[next] - 1
			if incoming[next] == 0 {
				queue = append(queue, next)
			}
		}
	}

	if len(ordered) != len(nodeIDs) {
		seen := make(map[string]bool, len(ordered))
		for _, id := range ordered {
			seen[id] = true
		}
		for _, id := range nodeIDs {
			if !seen[id] {
				ordered = append(ordered, id)
			}
		}
	}

	return ordered
}

